config/identity.go

```diff
@@ -0,0 +1,193 @@
+package config
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/kelseyhightower/envconfig"
+	crypto "github.com/libp2p/go-libp2p-crypto"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+const configKey = "cluster"
+
+// Identity defaults
+const (
+	DefaultConfigCrypto    = crypto.Ed25519
+	DefaultConfigKeyLength = -1
+)
+
+// Identity represents identity of a cluster peer for communication,
+// including the Consensus component.
+type Identity struct {
+	ID         peer.ID
+	PrivateKey crypto.PrivKey
+}
+
+// identityJSON represents a Cluster peer identity as it will look when it is
+// saved using JSON.
+type identityJSON struct {
+	ID         string `json:"id"`
+	PrivateKey string `json:"private_key"`
+}
+
+// NewIdentity generate a public-private keypair and returns a new Identity.
+func NewIdentity() (*Identity, error) {
+	// pid and private key generation --
+	priv, pub, err := crypto.GenerateKeyPair(
+		DefaultConfigCrypto,
+		DefaultConfigKeyLength)
+	if err != nil {
+		return nil, err
+	}
+	pid, err := peer.IDFromPublicKey(pub)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Identity{
+		ID:         pid,
+		PrivateKey: priv,
+	}, nil
+}
+
+// ConfigKey returns a human-readable string to identify
+// a cluster Identity.
+func (ident *Identity) ConfigKey() string {
+	return configKey
+}
+
+// SaveJSON saves the JSON representation of the Identity to
+// the given path.
+func (ident *Identity) SaveJSON(path string) error {
+	logger.Info("Saving configuration")
+
+	bs, err := ident.ToJSON()
+	if err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(path, bs, 0600)
+}
+
+// ToJSON generates a human-friendly version of Identity.
+func (ident *Identity) ToJSON() (raw []byte, err error) {
+	jID, err := ident.toIdentityJSON()
+	if err != nil {
+		return
+	}
+
+	raw, err = json.MarshalIndent(jID, "", "    ")
+	return
+}
+
+func (ident *Identity) toIdentityJSON() (jID *identityJSON, err error) {
+	// Multiaddress String() may panic
```

Comment:
This comment and the deferred `recover` appear to be copy-pasta

Reply:
my bad

Reply:
https://github.com/ipfs/ipfs-cluster/pull/766/commits/64390be061e4737aeb19a81066536720d291eea8